consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,394 @@
+# Specification for Celo IBFT protocol
+
+## Celo IBFT specification
+
+### High level overview
+
+The Celo IBFT protocol is a BFT (Byzantine Fault Tolerant) protocol that allows
+a group of participants to agree on the ordering of values by exchanging voting
+messages across a network. As long as less than 1/3rd of participants deviate
+from the protocol then the protocol should ensure that there is only one
+ordering of values that participants agree on and also that and that they can
+continue to agree on new values, i.e. they don't get stuck.
+
+Ensuring that only one ordering is agreed upon is referred to as 'safety' and
+ensuring that the participants can continue to agree on new values is referred
+to as 'liveness'.
+
+### System model
+
+We consider a group of `3f+1` participating computers (participants) that are
+able to communicate across a network. The group should be able to tolerate `f`
+failures without losing liveness or safety.
+
+All participants know the other participants in the system and it is assumed
+that all messages are signed in some way such that participants in the protocol
+know which participant a message came from and that messages cannot be forged.
+
```

Comment:
Similar to the reasoning for not mentioning block explicitly, from the perspective of consensus algorithm and if you think about it the [implementation](https://github.com/celo-org/celo-blockchain/blob/58804ca7236f60b19ed7fd3ec9d0fc3284621b00/consensus/istanbul/core/handler.go#L163-L182), when it receives a message it does not know, or need to know that the message was sent over the celo p2p network. Changing the network implementation has no effect on the consensus algorithm, and explicitly coupling the consensus to the p2p implementation means that if we were ever to change the network implementation we would need to go and modify the consensus spec to represent that, which is work that is avoidable.

I'm wondering if its not clear what I'm trying to specify here? If I were trying to specify the working of the entire collection of packages under `consensus/istanbul` then it would be pertinent to specify these details, but I'm only trying to specify at a conceptual level the algorithm used for achieving consensus.

Reply:
Ok, makes sense. Maybe add a paragraph clarifying this at the start of the doc?

Reply:
I've added a description up front for this.

https://github.com/celo-org/celo-blockchain/blob/5afaaa5bb877cb106d9dec1567df41dd8eee7668/consensus/istanbul/celo-ibft-spec.md?plain=1#L5-L15